plugins/synapticsmst/synapticsmst-device.c

```diff
@@ -586,139 +1094,242 @@ synapticsmst_device_write_firmware (SynapticsMSTDevice *device,
+		}
+	}
+	/* open device */
 	if (!synapticsmst_device_open (device, error)) {
 		g_prefix_error (error,
 				"can't open DP Aux node %s",
 				synapticsmst_device_get_aux_node (device));
 		return FALSE;
 	}
-
 	/* enable remote control */
 	if (!synapticsmst_device_enable_remote_control (device, error))
 		return FALSE;
 
-	/* erase SPI flash */
-	connection = synapticsmst_common_new (priv->fd, priv->layer, priv->rad);
-	if (synapticsmst_common_rc_set_command (connection,
-						UPDC_FLASH_ERASE,
-						2, 0, (guint8 *)&erase_code)) {
-		g_set_error_literal (error,
-				     G_IO_ERROR,
-				     G_IO_ERROR_INVALID_DATA,
-				     "can't erase flash");
-		return FALSE;
+	/* detect SPI quad mode & disable HDCP 2.2 */
+	if(isChipPanamera == TRUE)
+	{
+		/* Need to detect flash mode and ESM first ? */
+		/* disable flash Quad mode and ESM/HDCP2.2*/
+		for(guint8 i=0; i<16; i++){
+			dwData[i] = 0;	
+		}
+		rc = 0;
+		connection = synapticsmst_common_new (priv->fd, priv->layer, priv->rad);		
+		/* disable ESM first */
+		dwData[0] = 0x21;
+		rc = synapticsmst_common_rc_set_command (connection, UPDC_WRITE_TO_MEMORY,
+								4, (gint)0x2000fc, (guint8*)dwData);
+		g_usleep(200);// waiting for ESM exit
+		/* disable QUAD mode */ 
+		rc = synapticsmst_common_rc_get_command (connection, UPDC_READ_FROM_MEMORY,
+								((sizeof(dwData)/sizeof(dwData[0]))*4), (gint)0x200fc0, (guint8*)dwData);		
+		dwData[0] = 0x00;
+		rc = synapticsmst_common_rc_set_command (connection, UPDC_WRITE_TO_MEMORY,
+								4, (gint)0x200fc0, (guint8*)dwData);	
+		/* disable HDCP2.2 */
+		rc = synapticsmst_common_rc_get_command (connection, UPDC_READ_FROM_MEMORY,
+								4, (gint)0x200f90, (guint8*)dwData);		
+		dwData[ 0 ] = dwData[0] & (~BIT(2));
+		rc = synapticsmst_common_rc_set_command (connection, UPDC_WRITE_TO_MEMORY,
+								4, (gint)0x200f90, (guint8*)dwData);		
+		if(rc)								
+			g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA,
+				     "can't disable Quad & HDCP");		
+	}
+
+	if(isChipPanamera == TRUE){
+		//WP control
+		rc = 0;
+		//rc = synapticsmst_device_disable_flash_write_protect(device, TRUE);
+	}
+	else{
+		connection = synapticsmst_common_new (priv->fd, priv->layer, priv->rad);
+		if (synapticsmst_common_rc_set_command (connection,
+							UPDC_FLASH_ERASE,
+							2, 0, (guint8 *)&erase_code)) {
+			g_set_error_literal (error,
+						G_IO_ERROR,
+						G_IO_ERROR_INVALID_DATA,
+						"can't erase flash");
+			return FALSE;
+		}
 	}
 
 	/* update firmware */
-	write_loops = (payload_len / BLOCK_UNIT);
-	data_to_write = payload_len;
-	rc = 0;
-	offset = 0;
+	if(isChipPanamera == TRUE){
+		rc = 0;
+		rc = synapticsmst_device_update_ESM(device, payload_data, error);
+		if(rc == 0){								
+			g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA,
+				     "Update ESM fail");		
+			return FALSE;
+		}
+		g_usleep(100);
+		rc = 0;
+		rc = synapticsmst_device_update_Panamera_firmware(device, payload_len ,payload_data, error);
```

Comment:
When you make this function boolean and make sure all the `FALSE` conditions have set `GError` then your test can look like this:
```
if (!synapticsmst_device_update_panamera_firmware (device, payload_len ,payload_data, error)) {
	g_prefix_error (error, "Failed to update firmware: ");
	return FALSE;
}
```

Then when the error comes up to the console it will show up as a string of events that make it really easy to trace where the problem is.

Reply:
Thanks, will modify lately.